### DISCUSSION TITLE
Tab size is too wide when viewing code

### DISCUSSION THREAD

**Discussion by Levi-Lesches** [community/community#4893] [Other Feature Feedback, Questions, & Ideas] [answered]
Currently, tabs are set to be 8 spaces, which makes looking at tab-formatted code really difficult, and lines wrap everywhere. This is already a way to change the tab size when _editing_, but not while viewing. Simply having this option while viewing code would make a big difference. 
![image](https://user-images.githubusercontent.com/20747538/127908184-fc2e4039-72ee-4458-ba1f-48c14368080e.png)

I'm aware of the `.editorconfig` hack, but tab sizes are a per-user preference, and the whole point of using them is that you're _not_ forcing your preferences on everyone else. The `?ts=` hack works great, but not well-known or intuitive. Having a UI component that does the same thing would be very helpful. 

This issue has gotten significant attention at https://github.com/isaacs/github/issues/170, with almost 200 👍. (Make sure to _upvote_ this, not just add a 👍).

**Comment by Alhadis** [comment]
> Having a UI component that does the same thing would be very helpful.

And if that's too much to ask for, then please, **_please_**, add a repository-level setting to configure the default tab-width so incorrigible hacks [like this](https://github.com/Alhadis/Utils/commit/67cfd0571e1e3b3b3228c1e0988e0aa3cd914b0a) are no longer necessary. It defaults the whole purpose of semantic indentation.

I've been requesting this feature for years, even going so far as to include code samples to overcome my inability to submit a pull-request. The frustrating part is that this is a VERY simple feature to implement, and one that's been heavily requested time and time again by countless users. Listen to us, GitHub. Please/

**Reply by scunliffe** [reply]
Thanks for the `.editorconfig` hack info... while not ideal, this temporary solution is quite helpful!

**Reply by nhed** [reply]
Just tested and I don't think the editor config even works on a per-repo basis

**Comment by dipree** [comment, marked as answer]
Hey everyone 👋 good news, you can now define the tab size in the [Appearance settings](https://github.com/settings/appearance) of your user account. Code rendered on GitHub with tab indent will use your preferred tab size.

**Reply by lucasbasquerotto** [reply]
Great news!

**Reply by Levi-Lesches** [reply]
Amazing! For those who wanted a repo-wide setting, I'm going to mark this as the answer since the hallmark of tabs is letting everyone choose their own preference.

**Reply by Alhadis** [reply]
**Nitpick:** This shouldn't take precedence over the existing EditorConfig hack—authors who continue to enforce a specific tab-width probably have good reasons for doing so (generally involving stuff like tabular data and legacy codebases).

In other words, this should only take place in the *absence* of a configured tab-stop width.

**Reply by dipree** [reply]
@Alhadis thanks for noting. For accessibility reasons we decided: personal preference > editor config.

**Reply by Alhadis** [reply]
@dipree The situation's more complex than that. See [#5813](https://github.com/github/feedback/discussions/5813).

**Reply by Levi-Lesches** [reply]
I posted this there, but 
- the whole point of using tabs is to give the users control, not the repo
- if a repo is really unreadable, the user can switch their preference to the repo's style
- if a repo really wants to _force_ their style, they should use spaces

**Reply by lucasbasquerotto** [reply]
I agree with @Levi-Lesches here. I saw somewhere something that I agree: **Tabs for indentation, spaces for alignment**. If the project is yours, I advise to use spaces if you want alignment. If it's a third party project you might not have much to do about, except changing your tab size or cloning the repo and seeing the file with a custom tab size (vscode allows that, for example). I don't consider a very good idea the repository forcing the tab size to users.

**Reply by Alhadis** [reply]
> the whole point of using tabs is to give the users control, not the repo

I agree, but in cases where the author can't/won't indent their code semantically, users are left with unreadable source code unless they change a site-wide preference (which nobody wants to do for casual perusal).

> if a repo really wants to _force_ their style, they should use spaces

Some coding-styles mandate a fixed tab-width for some reason, most notably the [Linux Kernel](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#indentation).

**Full disclosure:** I've been advocating the [_tabs for indentation_](https://github.com/Alhadis/Coding-Style#1-use-tabs-for-indentation), [_spaces for alignment_](https://github.com/Alhadis/Coding-Style#2-dont-use-tabs-for-alignment) argument for many years, so I'm certainly not promoting or defending non-semantic tab use. However, I don't want the readability of existing projects to suffer because of this feature, either. Projects that have an opinionated `indent_size` in their EditorConfig will benefit from users submitting a PR to remove it (not just for GitHub's benefit, either).

**Reply by mrleblanc101** [reply]
Hi,
I'm pretty sure this is not 100% working as intended.
The tab-size is not respected when viewing rendered `README` (or any other Markdown file for that matter).
[Here](https://github.com/oliverfindl/vue-svg-inline-loader) is a README file that use Tabs in a codeblock.
My settings is set to 2 space but it's still rendered as 8 spaces.
![Capture d’écran, le 2022-06-29 à 13 14 40](https://user-images.githubusercontent.com/17392251/176496869-f0760aa3-e80a-4fc3-8724-f8ab98e08896.png)
![Capture d’écran, le 2022-06-29 à 13 16 27](https://user-images.githubusercontent.com/17392251/176497071-884349cf-fe37-487e-881f-e9a226e0a9a9.png)
Here is a comparaison with VS Code output with proof that it use Tabs and not Space.
<img width="469" alt="Capture d’écran, le 2022-06-29 à 13 18 03" src="https://user-images.githubusercontent.com/17392251/176497437-81bb3486-fc6e-4f6a-a96e-c40ecf5ecc3b.png">

**Reply by runeflobakk** [reply]
I also experience this. Tab-size is not respected when viewing rendered markdown. Which is a pity, because I prefer to use tabs because of accessibility concerns for documentation, and to allow people with different needs to adjust how they prefer the size of indentation to appear on their screens.

**Reply by Alhadis** [reply]
Injecting the following CSS into the page for logged-in users should solve this issue. Server-side, this might look something like:

```html+erb
<style>
.markdown-body .highlight pre,
.markdown-body pre {
	tab-size: <%= user.preferences.tab_size || 8 %>;
}
</style>
```

(Tailor the embedded Ruby command to whatever's used to retrieve a user's tab-size preference).

**Reply by uncenter** [reply]
Great that this is configurable but why does it still default to 8 spaces? Absolutely nobody wants or needs it that wide. Should be 4 spaces at the most.

**Reply by nhed** [reply]
> Great that this is configurable but why does it still default to 8 spaces? Absolutely nobody wants or needs it that wide. Should be 4 spaces at the most.

Linux kernel uses 8, git and Linux were written by Linux - maybe that's github's homage to Linus

My pref is 2 spaces but no one asked me :-)

**Reply by georgealways** [reply]
**Please consider making 4 characters the default.** 8 is too wide, and it's a very uncommon preference to use as a default. 

It's great that this setting is configurable, but I'd guess that most people don't know the option exists.

**Reply by Julieisbaka** [reply]
2 Spaces OP (though for certain languages 4 looks better like i feel like for json 4 looks better but idk)
